toqito/nonlocal_games/extended_nonlocal_game.py

```diff
@@ -247,173 +254,215 @@ def nonsignaling_value(self) -> float:
-                prev_win = lower_bound
-
-                # As the SDPs keep alternating, check if the winning probability
-                # becomes any higher. If so, replace with new best.
-                best = max(best, lower_bound)
-
-            best_lower_bound = max(best, best_lower_bound)
-
-        return best_lower_bound
-
-    def __optimize_alice(self, bob_povms) -> tuple[dict, float]:
+                    ket = random_u_mat[:, b_ans].reshape(-1, 1)
+                    bra = ket.conj().T
+                    bob_povms_np[y_ques, b_ans] = ket @ bra
+        else:
+            for y_ques in range(num_inputs_bob):
+                bob_povms_np[y_ques, 0] = np.eye(num_outputs_bob)
+                for b_other_ans in range(1, num_outputs_bob):
+                    bob_povms_np[y_ques, b_other_ans] = np.zeros((num_outputs_bob, num_outputs_bob))
+
+        prev_win_val = -float("inf")
+        current_best_lower_bound = -float("inf")
+
+        # print(f"Starting see-saw: max_steps={iters}, tol={tol}, seed={seed}, solver={solver}")
+
+        for step in range(iters):
+            # Pass solver and solver_params
+            opt_alice_rho_cvxpy_vars, problem_alice = self.__optimize_alice(bob_povms_np, solver, solver_params)
+
+            if (
+                opt_alice_rho_cvxpy_vars is None
+                or problem_alice.status not in [cvxpy.OPTIMAL, cvxpy.OPTIMAL_INACCURATE]
+                or problem_alice.value is None
+            ):
+                # print(f"Warning: Alice opt failed (status: {problem_alice.status})
+                # in step {step + 1}. Val: {problem_alice.value}")
+                return current_best_lower_bound if current_best_lower_bound > -float("inf") else 0.0
+
+            # Pass solver and solver_params
+            opt_bob_povm_cvxpy_vars, problem_bob = self.__optimize_bob(opt_alice_rho_cvxpy_vars, solver, solver_params)
+
+            if (
+                opt_bob_povm_cvxpy_vars is None
+                or problem_bob.status not in [cvxpy.OPTIMAL, cvxpy.OPTIMAL_INACCURATE]
+                or problem_bob.value is None
+            ):
+                # print(f"Warning: Bob opt failed (status: {problem_bob.status})
+                # in step {step + 1}. Val: {problem_bob.value}")
+                return current_best_lower_bound if current_best_lower_bound > -float("inf") else 0.0
+
+            current_win_val = problem_bob.value
+
+            current_best_lower_bound = max(current_best_lower_bound, current_win_val)
+
+            improvement = abs(current_win_val - prev_win_val)
+
+            if step < iters - 1:
+                for y_idx in range(self.num_bob_in):
+                    for b_idx in range(self.num_bob_out):
+                        bob_povms_np[y_idx, b_idx] = opt_bob_povm_cvxpy_vars[y_idx, b_idx].value
+                    # else:
+                    # print(f"Warning: Bob POVM var ({y_idx},{b_idx}) value is None in step {step+1}.")
+                    # break
```

Comment:
I am not too familiar with the math here. Is there a use case for when this value can ever be `None`?